packages/box/box.ts

```diff
@@ -19,50 +21,132 @@ const defaultPath = "https://github.com:trufflesuite/truffle-init-default";
  *   - git@github.com:<org>/<repo>[#branch]
  *   - path to local folder (absolute, relative or ~/home)
  */
-const normalizeSourcePath = (url = defaultPath) => {
+export const normalizeSourcePath = (url = defaultPath) => {
+  // Process filepath resolution
+  //
   if (url.startsWith(".") || url.startsWith("/") || url.startsWith("~")) {
+    debug({ in: url, out: path.normalize(url) });
     return path.resolve(path.normalize(url));
   }
-  // remove the .git from the repo specifier
-  if (url.includes(".git")) {
-    url = url.replace(/.git$/, "");
-    url = url.replace(/.git#/, "#");
-    url = url.replace(/.git:/, ":");
-  }
 
-  // rewrite https://github.com/truffle-box/metacoin format in
-  //         https://github.com:truffle-box/metacoin format
-  if (url.match(/.com\//)) {
-    url = url.replace(/.com\//, ".com:");
-  }
+  // preprocess to reduce regex complexity
+  // `https` is not case sensitiv, unlike `git`
+  url = url.replace(/^https/i, "https");
 
-  // full URL already
-  if (url.includes("://")) {
-    return url;
-  }
+  // branch should not end with slash
+  const invalidBranch = /\/$/;
 
-  if (url.includes("git@")) {
-    return url.replace("git@", "https://");
-  }
+  // process https? or git prefixed input
+  //
+  if (/^(https?|git)/i.test(url)) {
+    // This regular expression uses named capture groups to parse input. The
+    // format is (?<the-name>the-regex)
+    //
+    // \w, the word meta character is a member of [A-Za-z0-9_]. all letters,
+    // digits and the underscore. Note \w has to be \\w to escape the backslash
+    // in a string literal.
+    //
+    const protocolRex = new RegExp([
+      // match either `htps://` or `git@`
+      "(?<protocol>(https://|git@))",
 
-  if (url.split("/").length === 2) {
-    // `org/repo`
-    return `https://github.com:${url}`;
-  }
+      // service is 1 or many (word, dot or dash)
+      "(?<service>[\\w.-]+)",
+
+      // match either `/` or `:`
+      "(/|:)",
+
+      // org is 1 or many (word, dot or dash)
+      "(?<org>[\\w.-]+)",
+
+      "/",
 
-  if (!url.includes("/")) {
-    // repo name only
-    if (!url.includes("-box")) {
-      // check for branch
-      if (!url.includes("#")) {
-        url = `${url}-box`;
-      } else {
-        const index = url.indexOf("#");
-        url = `${url.substr(0, index)}-box${url.substr(index)}`;
+      // repo is 1 or many (word, dot or dash)
+      "(?<repo>[\\w.-]+)",
+
+      // branch is 1 or many (word, dot or dash) and can be optional
+      "(?<branch>#[\\w./-]+)?",
+
+      // the input string must be consumed fully at this point to match
+      "$",
+    ].join(""));
+
+    const match = url.match(protocolRex);
+    if (match) {
+      const { groups: G } = match;
+      const branch = G["branch"] ? `${G["branch"]}` : "";
```

Comment:
Does `${G["branch"]}` need the \`${..}\`? Why not just `G["branch"]`?

Reply:
yes, that's better!